src/libcollections/str.rs

```diff
@@ -532,48 +593,129 @@ impl str {
+    /// As a string slice consists of valid UTF-8, we can iterate through a
+    /// string slice by `char`. This method returns such an iterator.
     ///
-    /// For iteration over human-readable characters a grapheme cluster iterator
-    /// may be more appropriate. See the [unicode-segmentation crate][1].
+    /// # Examples
     ///
-    /// [1]: https://crates.io/crates/unicode-segmentation
+    /// Basic usage:
     ///
-    /// # Examples
+    /// ```
+    /// let word = "goodbye";
+    ///
+    /// let count = word.chars().count();
+    /// assert_eq!(7, count);
+    ///
+    /// let mut chars = word.chars();
+    ///
+    /// assert_eq!(Some('g'), chars.next());
+    /// assert_eq!(Some('o'), chars.next());
+    /// assert_eq!(Some('o'), chars.next());
+    /// assert_eq!(Some('d'), chars.next());
+    /// assert_eq!(Some('b'), chars.next());
+    /// assert_eq!(Some('y'), chars.next());
+    /// assert_eq!(Some('e'), chars.next());
+    ///
+    /// assert_eq!(None, chars.next());
+    /// ```
+    ///
+    /// Remember, `char`s may not match your human intuition about characters:
     ///
     /// ```
-    /// let v: Vec<char> = "ASCII żółć 🇨🇭 한".chars().collect();
+    /// let y = "y̆";
     ///
-    /// assert_eq!(v, ['A', 'S', 'C', 'I', 'I', ' ',
-    ///     'z', '\u{307}', 'o', '\u{301}', 'ł', 'c', '\u{301}', ' ',
-    ///     '\u{1f1e8}', '\u{1f1ed}', ' ', '한']);
+    /// let mut chars = y.chars();
+    ///
+    /// assert_eq!(Some('y'), chars.next()); // not 'y̆'
+    /// assert_eq!(Some('\u{0306}'), chars.next());
+    ///
+    /// assert_eq!(None, chars.next());
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     #[inline]
     pub fn chars(&self) -> Chars {
         core_str::StrExt::chars(self)
     }
-
-    /// An iterator over the `char`s of `self` and their byte offsets.
+    /// Returns an iterator over the `char`s of a string slice, and their
+    /// positions.
+    ///
+    /// As a string slice consists of valid UTF-8, we can iterate through a
+    /// string slice by `char`. This method returns an iterator of both
+    /// these `char`s, as well as their positions, in bytes.
+    ///
+    /// The iterator yields tuples. The position is first, the `char` is
+    /// second.
     ///
     /// # Examples
     ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// let word = "goodbye";
+    ///
+    /// let count = word.char_indices().count();
+    /// assert_eq!(7, count);
+    ///
+    /// let mut char_indices = word.char_indices();
+    ///
+    /// assert_eq!(Some((0, 'g')), char_indices.next());
+    /// assert_eq!(Some((1, 'o')), char_indices.next());
+    /// assert_eq!(Some((2, 'o')), char_indices.next());
+    /// assert_eq!(Some((3, 'd')), char_indices.next());
+    /// assert_eq!(Some((4, 'b')), char_indices.next());
+    /// assert_eq!(Some((5, 'y')), char_indices.next());
+    /// assert_eq!(Some((6, 'e')), char_indices.next());
```

Comment:
This example looks pretty long winded. Maybe selecting some shorter string for `word` would be better?